src/packagedcode/phpcomposer.py

```diff
@@ -0,0 +1,403 @@
+    def recognize(cls, location):
+        return parse(location)
+
+
+def is_phpcomposer_json(location):
+    return (filetype.is_file(location)
+            and fileutils.file_name(location).lower() == 'composer.json')
+
+
+def parse(location):
+    """
+    Return a Package object from a composer.json file or None.
+    """
+    if not is_phpcomposer_json(location):
+        return
+    # mapping of top level composer.json items to the Package object field name
+    plain_fields = OrderedDict([
+        ('name', 'name'),
+        ('description', 'summary'),
+        ('keywords', 'keywords'),
+        ('homepage', 'homepage_url'),
+    ])
+
+    # mapping of top level package.json items to a function accepting as arguments
+    # the package.json element value and returning an iterable of key, values Package Object to update
+    field_mappers = OrderedDict([
+        ('authors', author_mapper),
+        ('license', licensing_mapper),
+        ('require', dependencies_mapper),
+        ('require-dev', dev_dependencies_mapper),
+        ('repositories', repository_mapper),
+        ('support', support_mapper),
+    ])
+
+    with codecs.open(location, encoding='utf-8') as loc:
+        data = json.load(loc, object_pairs_hook=OrderedDict)
+
+    if not data.get('name') or not data.get('description'):
+        # a package.json without name and description is not a usable PHP composer package
+        return
+
+    package = PHPComposerPackage()
+    # a composer.json is at the root of a PHP composer package
+    base_dir = fileutils.parent_directory(location)
+    package.location = base_dir
+    # for now we only recognize a package.json, not a node_modules directory yet
+    package.metafile_locations = [location]
+    package.version = data.get('version')
+    for source, target in plain_fields.items():
+        value = data.get(source)
+        if value:
+            if isinstance(value, basestring):
+                value = value.strip()
+            if value:
+                setattr(package, target, value)
+
+    for source, func in field_mappers.items():
+        logger.debug('parse: %(source)r, %(func)r' % locals())
+        value = data.get(source)
+        if value:
+            if isinstance(value, basestring):
+                value = value.strip()
+            if value:
+                func(value, package)
+    return package
+
+
+def licensing_mapper(licenses, package):
+    """
+    Update package licensing and return package.
+    Licensing data structure has evolved over time and is a tad messy.
+    https://getcomposer.org/doc/04-schema.md#license
+    licenses is either:
+    - a string with:
+     - an SPDX id or expression {  "license": "(LGPL-2.1 or GPL-3.0+)" }
+    - array:
+        "license": [
+           "LGPL-2.1",
+           "GPL-3.0+"
+        ]
+        """
+    if not licenses:
+        return package
+
+    if isinstance(licenses, basestring):
+        if '(' in licenses and ')' in license and 'or' in license:
```

Comment:
Also at this stage, let's not try to parse license expressions yet. I reckon this is a tad odd, but we would need to do this with https://github.com/nexB/license-expression/ to be correct.